doc/sfputil/read_write_eeprom_by_page.md

```diff
@@ -0,0 +1,191 @@
+
+The existing API `sfp.read_eeprom` and `sfp.write_eeprom` accept "overall offset" as parameter. They have no concept of page. If user wants to use it, user has to convert page and offset to "overall offset" manually. Different cable types use different covert method according to the standard. So, it is not user friendly to provide such API directly to user. sonic-utilities shall provide the function to translate from page offset to overall offset.
+
+##### CMIS validation
+
+Passive cable:
+- Valid page: 0
+- Valid offset: 0-255
+
+Active cable:
+- Valid page: 0-255
+= Valid offset: page 0 (0-255), other (128-255)
+
+For active cable, there is no "perfect" page validation as it is too complicated. User is responsible to make sure the page existence according to cable user manual.
+
+Example:
+```
+sfputil read-eeprom Ethernet0 0 255 1               # valid
+sfputil read-eeprom Ethernet0 0 255 2               # invalid size, out of range, 255+2=257 is not a valid offset
+sfputil read-eeprom Ethernet0 0 256 1               # invalid offset 256 for page 0, must be in range [0, 255]
+sfputil read-eeprom Ethernet0 1 0 1                 # invalid offset 0 for page 1, must be >=128
+```
+
+##### sff8436 and sff8636 validation
+
+Passive cable:
+- Valid page: 0
+- Valid offset: 0-255
+
+Active cable:
+- Valid page: 0-255
+- Valid offset: page 0 (0-255), other (128-255)
+
+For active cable, there is no "perfect" page validation as it is too complicated. User is responsible to make sure the page existence according to cable user manual.
+
+```
+sfputil write-eeprom Ethernet0 0 255 ff            # valid
+sfputil write-eeprom Ethernet0 0 255 ff00          # invalid size, out of range, 255+2=257 is not a valid offset
+sfputil write-eeprom Ethernet0 0 256 ff            # invalid offset 256 for page 0, must be in range [0, 255]
+sfputil write-eeprom Ethernet0 1 0 ff              # invalid offset 0 for page 1, must be >=128
+```
+
+##### sff8472 validation
+
+Passive cable:
+- Valid wire address: [A0h] (case insensitive)
+- Valid offset: A0h (0-128)
+
+Active cable:
+- Valid wire address: [A0h, A2h] (case insensitive)
+- Valid offset: A0h (0-255), A2h (0-255)
+
+```
+sfputil read-eeprom Ethernet0 0 0 1 --wire-addr a0h               # valid
+sfputil read-eeprom Ethernet0 0 0 2 --wire-addr A0h               # invalid size, out of range, 255+2=257 is not a valid offset
+sfputil read-eeprom Ethernet0 1 0 1 --wire-addr a2h               # invalid offset 256 for page 0, must be in range [0, 255]
+sfputil read-eeprom Ethernet0 1 0 1 --wire-addr a0h               # invalid offset 0 for page 1, must be >=128
+```
+
+### SAI API
+
+N/A
+
+### Configuration and management
+
+#### Manifest (if the feature is an Application Extension)
+
+N/A
+
+#### CLI/YANG model Enhancements
+
+##### sfputil read-eeprom
+
+```
+admin@sonic:~$ sfputil read-eeprom --help
+Usage: sfputil read-eeprom [OPTIONS] <port_name> <page> <offset> <size>

```

Comment:
@Junchao-Mellanox For clarity can we have explicit CLI option
-p <PORT> mandatory 

--page <PAGE> optional. How to specify lower page and upper page 0

--o <Offset>

Reply:
"How to specify lower page and upper page 0"

lower page 0 and upper page 0 are all treated as page 0. lower page 0 starts from offset 0, upper page 0 starts from offset 128.

Reply:
@Junchao-Mellanox if page=0 and offset > 127 then its lower page
if page=0 and offset <= 127 its upper page. Cann't you use this logic?

Reply:
@Junchao-Mellanox also, please provide mandatory option like this

```
sfputil read-eeprom --help
-p port name
-n page no
-o offset
-s size
```

This way when user specifies the option like:
```
spfuilt read-eeprom -p EthernetXX -n 0 -o 240 -s 10
```

Is much more read able than below
```
spfuilt read-eeprom EthernetXX   0   240  10
```

Reply:
Fixed